src/librbd/api/Group.cc

```diff
@@ -26,20 +30,280 @@ using librados::Rados;
+  }
+  string group_header_oid = util::group_header_name(group_id);
+
+  ldout(cct, 20) << "listing images in group name "
+		 << group_name << " group id " << group_header_oid << dendl;
+  image_ids->clear();
+
+  const int max_read = 1024;
+  cls::rbd::GroupImageSpec start_last;
+  do {
+    std::vector<cls::rbd::GroupImageStatus> image_ids_page;
+
+    r = cls_client::group_image_list(&group_ioctx, group_header_oid,
+				     start_last, max_read, &image_ids_page);
+
+    if (r < 0) {
+      lderr(cct) << "error reading image list from consistency group: "
+	<< cpp_strerror(-r) << dendl;
+      return r;
+    }
+    image_ids->insert(image_ids->end(),
+		     image_ids_page.begin(), image_ids_page.end());
+
+    if (image_ids_page.size() > 0)
+      start_last = image_ids_page.rbegin()->spec;
+
+    r = image_ids_page.size();
+  } while (r == max_read);
+
+  return 0;
+}
+
+static int group_image_remove(librados::IoCtx& group_ioctx, string group_id,
+			      librados::IoCtx& image_ioctx, string image_id)
+{
+  CephContext *cct = (CephContext *)group_ioctx.cct();
+
+  string group_header_oid = util::group_header_name(group_id);
+
+  string image_header_oid = util::header_name(image_id);
+
+  ldout(cct, 20) << "removing image " << image_id
+		 << " image id " << image_header_oid << dendl;
+
+  cls::rbd::GroupSpec group_spec(group_id, group_ioctx.get_id());
+
+  cls::rbd::GroupImageStatus incomplete_st(image_id, image_ioctx.get_id(),
+				cls::rbd::GROUP_IMAGE_LINK_STATE_INCOMPLETE);
+
+  cls::rbd::GroupImageSpec spec(image_id, image_ioctx.get_id());
+
+  int r = cls_client::group_image_set(&group_ioctx, group_header_oid,
+				      incomplete_st);
+
+  if (r < 0) {
+    lderr(cct) << "couldn't put image into removing state: "
+	       << cpp_strerror(-r) << dendl;
+    return r;
+  }
+
+  r = cls_client::image_remove_group(&image_ioctx, image_header_oid,
+				     group_spec);
+  if ((r < 0) && (r != -ENOENT)) {
+    lderr(cct) << "couldn't remove group reference from image"
+	       << cpp_strerror(-r) << dendl;
+    return r;
+  }
+
+  r = cls_client::group_image_remove(&group_ioctx, group_header_oid, spec);
+  if (r < 0) {
+    lderr(cct) << "couldn't remove image from group"
+	       << cpp_strerror(-r) << dendl;
+    return r;
+  }
+
+  return 0;
+}
+
+static int group_snap_remove_by_record(librados::IoCtx& group_ioctx,
+    const cls::rbd::GroupSnapshot& group_snap, const std::string& group_id,
+    const std::string& group_header_oid) {
+
+  CephContext *cct = (CephContext *)group_ioctx.cct();
+  librados::Rados rados(group_ioctx);
+  std::vector<C_SaferCond*> on_finishes;
+  int r, ret_code;
+
+  std::vector<librbd::IoCtx*> io_ctxs;
+  std::vector<librbd::ImageCtx*> ictxs;
+
+  std::string image_snap_name;
+
+  cls::rbd::SnapshotNamespace ne;
+
+  ldout(cct, 20) << "Removing snapshots" << dendl;
+  int snap_count = group_snap.snaps.size();
+
+  for (int i = 0; i < snap_count; ++i) {
+    librbd::IoCtx* image_io_ctx = new librbd::IoCtx;
```

Comment:
Unnecessary to allocate on the heap -- it internally reference counts (plus ```ImageCtx``` just creates its own immediately)